stumpy/stump.py

```diff
@@ -400,46 +453,72 @@ def _stump(
             thread_idx,
             ρ,
             I,
+            ρL,
+            IL,
+            ρR,
+            IR,
             ignore_trivial,
+            k,
         )
 
     # Reduction of results from all threads
     for thread_idx in range(1, n_threads):
         for i in prange(l):
-            if ρ[0, i, 0] < ρ[thread_idx, i, 0]:
-                ρ[0, i, 0] = ρ[thread_idx, i, 0]
-                I[0, i, 0] = I[thread_idx, i, 0]
-            # left pearson correlation and left matrix profile indices
-            if ρ[0, i, 1] < ρ[thread_idx, i, 1]:
-                ρ[0, i, 1] = ρ[thread_idx, i, 1]
-                I[0, i, 1] = I[thread_idx, i, 1]
-            # right pearson correlation and right matrix profile indices
-            if ρ[0, i, 2] < ρ[thread_idx, i, 2]:
-                ρ[0, i, 2] = ρ[thread_idx, i, 2]
-                I[0, i, 2] = I[thread_idx, i, 2]
-
-    # Convert pearson correlations to distances
-    p_norm = np.abs(2 * m * (1 - ρ[0, :, :]))
+            # top-k
+            for j in range(
+                k - 1, -1, -1
+            ):  # reverse iteration to preserve order in ties
+                if ρ[0, i, 0] < ρ[thread_idx, i, j]:
+                    idx = np.searchsorted(ρ[0, i], ρ[thread_idx, i, j])
+                    ρ[0, i, : idx - 1] = ρ[0, i, 1:idx]
+                    ρ[0, i, idx - 1] = ρ[thread_idx, i, j]
+
+                    I[0, i, : idx - 1] = I[0, i, 1:idx]
+                    I[0, i, idx - 1] = I[thread_idx, i, j]
+
+            if ρL[0, i] < ρL[thread_idx, i]:
+                ρL[0, i] = ρL[thread_idx, i]
+                IL[0, i] = IL[thread_idx, i]
+
+            if ρR[0, i] < ρR[thread_idx, i]:
+                ρR[0, i] = ρR[thread_idx, i]
+                IR[0, i] = IR[thread_idx, i]
+
+    # Convert top-k pearson correlations to distances. The arrays ρ (and so I) should
+    # be reversed since ρ is in ascending order.
+    p_norm = np.abs(2 * m * (1 - ρ[0, :, ::-1]))
+    I = I[0, :, ::-1]
+
+    # Convert top-1 left/right pearson correlations to distances.
+    p_norm_L = np.abs(2 * m * (1 - ρL[0, :]))
+    p_norm_R = np.abs(2 * m * (1 - ρR[0, :]))
+
     for i in prange(p_norm.shape[0]):
-        if p_norm[i, 0] < config.STUMPY_P_NORM_THRESHOLD:
-            p_norm[i, 0] = 0.0
-        if p_norm[i, 1] < config.STUMPY_P_NORM_THRESHOLD:
-            p_norm[i, 1] = 0.0
-        if p_norm[i, 2] < config.STUMPY_P_NORM_THRESHOLD:
-            p_norm[i, 2] = 0.0
+        for j in prange(p_norm.shape[1]):
```

Comment:
> Loop serialization occurs when any number of prange driven loops are present inside another prange driven loop. In this case the outermost of all the prange loops executes in parallel and any inner prange loops (nested or otherwise) are treated as standard range based loops. Essentially, nested parallelism does not occur.

I read it in the document. I think this was your reason... You really LOVE Numba :)